core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -0,0 +1,341 @@
+     */
+    private final AtomicBoolean emitting = new AtomicBoolean(false);
+
+    private SplittingTransformer(AsyncResponseTransformer<ResponseT, ResultT> upstreamResponseTransformer,
+                                 Long bufferSize,
+                                 CompletableFuture<ResultT> returnFuture) {
+        this.upstreamResponseTransformer = Validate.paramNotNull(upstreamResponseTransformer, "asyncRequestBody");
+        this.returnFuture = Validate.paramNotNull(returnFuture, "returnFuture");
+        this.maximumBufferSize = Validate.notNull(bufferSize, "bufferSize");
+    }
+
+    /**
+     * @param downstreamSubscriber the {@link Subscriber} to the individual AsyncResponseTransformer
+     */
+    @Override
+    public void subscribe(Subscriber<? super AsyncResponseTransformer<ResponseT, ResponseT>> downstreamSubscriber) {
+        if (downstreamSubscriber == null) {
+            throw new NullPointerException("downstreamSubscriber must not be null");
+        }
+        this.downstreamSubscriber = downstreamSubscriber;
+        downstreamSubscriber.onSubscribe(new DownstreamSubscription());
+    }
+
+    /**
+     * The subscription implementation for the subscriber to this SplittingTransformer.
+     */
+    private final class DownstreamSubscription implements Subscription {
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                downstreamSubscriber.onError(new IllegalArgumentException("Amount requested must be positive"));
+                return;
+            }
+            if (isCancelled.get()) {
+                return;
+            }
+            long newDemand = outstandingDemand.updateAndGet(current -> {
+                if (Long.MAX_VALUE - current < n) {
+                    return Long.MAX_VALUE;
+                }
+                return current + n;
+            });
+            log.trace(() -> String.format("new outstanding demand: %s", newDemand));
+            emit();
+        }
+
+        @Override
+        public void cancel() {
+            if (isCancelled.compareAndSet(false, true)) {
+                log.trace(() -> "Cancelling splitting transformer");
+                publisherToUpstream.complete();
+                downstreamSubscriber = null;
+            }
+        }
+    }
+
+    private void emit() {
+        do {
+            if (!emitting.compareAndSet(false, true)) {
+                return;
+            }
+            try {
+                if (isCancelled.get()) {
+                    return;
+                }
+                if (outstandingDemand.get() > 0) {
+                    outstandingDemand.decrementAndGet();
+                    downstreamSubscriber.onNext(new IndividualTransformer());
+                }
+            } finally {
+                emitting.compareAndSet(true, false);
+            }
+        } while (outstandingDemand.get() > 0);
+    }
```

Comment:
It seems we are delivering one item at a time, suggesting the following for micro performance optimization. 

```
    private void emit() {
        do {
            if (!emitting.compareAndSet(false, true)) {
                return;
            }
            try {
                if (doEmit()) {
                    return;
                }

            } finally {
                emitting.compareAndSet(true, false);
            }
        } while (outstandingDemand.get() > 0);
    }

    private boolean doEmit() {
        long demand = outstandingDemand.get();

        while (demand > 0) {
            if (isCancelled.get()) {
                return true;
            }
            if (outstandingDemand.get() > 0) {
                demand = outstandingDemand.decrementAndGet();
                downstreamSubscriber.onNext(new IndividualTransformer());
            }
        }
        return false;
    }
```